Plugins/Public/alley/PlayerRestrictions.cpp

```diff
@@ -997,62 +1023,48 @@ bool ExecuteCommandString_Callback(CCmds* cmds, const wstring &wscCmd)
 {
 	returncode = DEFAULT_RETURNCODE;
-	if (iDmgToSpaceID && dmg->get_inflictor_id() && dmg->is_inflictor_a_player())
-	{
-		uint client = HkGetClientIDByShip(iDmgToSpaceID);
-		if (client)
-		{
-			uint ShootingClient = dmg->get_inflictor_owner_player();
-			Archetype::Ship* TheShipArch = Archetype::GetShip(Players[ShootingClient].iShipArchetype);
 
-			if (TheShipArch->iShipClass == 19)
-			{
-				float curr, max;
-				pub::SpaceObj::GetHealth(iDmgToSpaceID, curr, max);
-				float expecteddmg = (float)1;
-				float projecteddamage = curr - damage;
+	if (iDmgMunitionID != repairMunitionId
+		|| !iDmgTo
+		|| !dmg->is_inflictor_a_player()
+		|| p1 == 65521)
+		return;
 
-				//If a repair gun heals the ship but this doesn't show up, it's because it's hitting the shield.
-				//HkMsgU(L"DEBUG: damage by repair ship, is it healing?");					
-				//PrintUserCmdText(client, L"Projected damage: %f", projecteddamage);
+	if (!iDmgToSpaceID)
+	{
+		pub::Player::GetShip(iDmgTo, iDmgToSpaceID);
+	}
 
-				if ((projecteddamage <= 1) && (projecteddamage > 0))
-				{
-					//Handle the healing.
-					returncode = SKIPPLUGINS_NOFUNCTIONCALL;
+	float curr, maxHP;
+	pub::SpaceObj::GetHealth(iDmgToSpaceID, curr, maxHP);
 
-					Archetype::Ship* TheShipArchHealed = Archetype::GetShip(Players[client].iShipArchetype);
-					float amounttoheal = curr;
+	Archetype::Ship* TheShipArchHealed = Archetype::GetShip(Players[iDmgTo].iShipArchetype);
 
-					if (healingMultipliers.find(TheShipArchHealed->iShipClass) == healingMultipliers.end())
-					{
-						return;
-					}
-					// no need to check healingAdditions here as it is set at the same time as healingMultipliers
-					amounttoheal = max / 100 * healingMultipliers[TheShipArchHealed->iShipClass] + healingAdditions[TheShipArchHealed->iShipClass];
+	const auto& healingData = healingMultipliers.find(TheShipArchHealed->iShipClass);
+	if (healingData == healingMultipliers.end())
+		return;
 
-					float testhealth = curr + amounttoheal;
+	const HEALING_DATA& healing = healingData->second;
 
+	if (curr / maxHP >= healing.maxHeal)
+		return;
 
-					if (testhealth > max)
-					{
-						//HkMsgU(L"DEBUG: Health would be superior to max");
-						dmg->add_damage_entry(1, max, (DamageEntry::SubObjFate)0);
-						return;
-					}
-					else
-					{
-						//HkMsgU(L"DEBUG: Health less max");
-						dmg->add_damage_entry(1, testhealth, (DamageEntry::SubObjFate)0);
-						return;
-					}
-				}
-				//else do nothing, means it isn't a healing call.
-			}
-		}
+	float healedHP = (((healing.healingMultiplier / 100) + (curr / maxHP)) * maxHP) + healing.healingStatic;
+	float newHP = min(maxHP * healing.maxHeal, healedHP);
+
+	if (p1 == 1) 
+	{
+		damage = newHP;
+	}
+	else
+	{
+		iDmgTo = 0;
+		iDmgMunitionID = 0;
+		returncode = SKIPPLUGINS_NOFUNCTIONCALL;
+		dmg->add_damage_entry(1, newHP, static_cast<DamageEntry::SubObjFate>(0));
```

Comment:
Adding damage entry that will heal the main hull. 1 stands for hull internal id. NewHP is self-explanatory. Fate 0 means alive.

Reply:
This should be a comment in the code :)